Transact

hallie opened the pull request.

hallie committed:
adding support for transactions

pyi files

initial pass at chosen approach. add tests next

remove pyi changes

remove unused methods

pass args and kwargs instead

unit tests

removing unused imports

returning models from get response

missed a spot

adding transactions to base Connection

fixing tests

fixing tests and adding more

making Transaction of type object

more tests

more tests for models and base connection

adding integration tests

debug

adding service name

commenting out integration

bumping botocore v

adding host and region

handling case where response is already a list

adding some types

attempting to change how get method returns

fixing some tests

fixing more tests

removing contextmanager

helper function for mocking transaction results and kwargs for region and host

mocks are weird

removing kwargs

forgot to run unit tests ahead of push

adding more integration tests

forgot to change hash key

might have to choose an alternate approach for gets

removing yield and adding more integration tests

seeing what results look like

debugging key error

dang

dang, take 2

dang, take 3--sometimes it really just be like that

was that a race condition

no its because the host and region were missing...

or was it..

the internal server error was being caused by the update + action method

restructuring tests

more tests, and getting ddb url from env

addressing some nits and moving test setup to setup_class

debugging connection in tests

test remove mark

testing user table exists

host url in meta

more debugging

create tables outside of class

making these stand alone functions instead of a class

sleepin

not sleepin

more debugging

more debugging

updating GET interface, fixing tests

attempting to ensure resource exists

attempting to ensure resource exists

making all integration tests sequential insteas of function based

testing changing the region

fixing tests

undoing boto version change

betamoo commented on pynamodb/connection/base.py:
```diff
@@ -21,19 +21,17 @@
     BATCH_WRITE_ITEM, CONSISTENT_READ, ATTR_VALUE_LIST, DESCRIBE_TABLE, KEY_CONDITION_EXPRESSION,
     BATCH_GET_ITEM, DELETE_REQUEST, SELECT_VALUES, RETURN_VALUES, REQUEST_ITEMS, ATTR_UPDATES,
-    PROJECTION_EXPRESSION, SERVICE_NAME, DELETE_ITEM, PUT_REQUEST, UPDATE_ITEM, SCAN_FILTER, TABLE_NAME,
+    PROJECTION_EXPRESSION, SERVICE_NAME, DELETE_ITEM, PUT_REQUEST, UPDATE_ITEM, TABLE_NAME,
```
what's happening with the `SCAN_FILTER` ?

hallie replied:
someone deleted its usage from this file and didn't remove it from the imports

garrettheel committed:
Bypass most unnecessary botocore logic surrounding auth (#617)

Commit:
Add type hint for last_evaluated_key parameter (#624)

ikonst committed:
Document last_evaluated_key in index queries (#625)

ikonst committed:
Complete type stubs of pynamodb.pagination (#623)

leobarcellos committed:
Allow specifying billing mode when creating tables

hallie committed:
fixing mypys

hallie committed:
bump test coverage

hallie committed:
moving constants, frixing mypy, adding context

hallie committed:
apparently both defs need to be overloaded

hallie committed:
removing null proxy models

hallie committed:
undoing mypy overloading

hallie committed:
addressing some comments

hallie committed:
writing tests for _update_item_with_raw_data, making private, and adding mypy

hallie committed:
making transaction logic more like batching

hallie committed:
adding stubs for transaction

hallie committed:
fixing stub

hallie committed:
removing proxy model refresh logic from TransactWrite

hallie committed:
removing client token validation and fixing stubs

hallie committed:
returning model promises instead of models

hallie committed:
fixing promise

hallie committed:
debug

hallie committed:
passing class instead of model

garrettheel committed:
Remove deprecated conditional APIs (#632)

hallie committed:
Merge branch 'master' into transact

hallie committed:
fixing tests

hallie committed:
new interface who dis

hallie committed:
does .update work now?

hallie committed:
update test to be more practical

techdragon committed:
Add type stubs for Billing Mode constants (#626)

jpinner-lyft committed:
Drop python 2.6 compatibility code (#635)

jpinner-lyft committed:
Remove MapAttributeMeta (#638)

ikonst committed:
Fix BillingMode compatibility with dynamodb-local (#637)

dynamodb-local doesn't support the BillingMode property when creating tables. For “provisioning” billing mode, it's safe to omit since it's the default.

jpinner-lyft committed:
Replace Model._get_meta_data with hash and range key attributes (#639)

hallie committed:
fix mypy

hallie committed:
Merge branch 'master' into transact

hallie committed:
simplify operation kwargs

simplify operation kwargs

jpinner-lyft committed:
Remove LegacyBooleanAttribute (#643)

ikonst committed:
Require botocore>=1.12.54 (#648)

BillingMode was introduced in botocore 1.12.54, so we'll set it as the minimum version and partially undo #637.

ikonst committed:
unpin Python 2.6 dependencies (#646)

hallie committed:
Merge branch 'master' into transact

conjmurph committed:
Better exception message when index does not exist (#641)

`Connection.query` raises `ValueError("No hash key attribute for index: {}".format(index_name))` both for missing hash key in an existing index, and for a non-existent index. This improves the exception message in the latter case.

betamoo commented on README.rst:
```diff
@@ -108,25 +108,19 @@ first name begins with 'D':
-
-    # Deprecated, use UserModel.email=="djohn@company.org" instead
-    for user in UserModel.query("John", email__eq="djohn@company.org"):
+    for user in UserModel.query("Denver", UserModel.email=="djohn@company.org"):
```
this is the same example as above (it was to show the difference between `email__eq ` and `UserModel.email==`)
so it is redundant.

Was there a specific reason to remove the second example?

betamoo replied:
if we are just depracteing the old way we just should remove the second example altogether

hallie replied:
I didn't make this readme change :0

hallie replied:
You should look at the pr under pynamodb/Pynamodb instead of the one under Lyft (the Lyft repo isn't up to date with the latest pynamodb master, so a lot of these changes aren't actually mine)

betamoo replied:
alright will review the other one

hallie committed:
Update pynamodb/connection/transactions.py

Co-Authored-By: Ilya Konstantinov <ilya.konstantinov@gmail.com>

hallie committed:
Update pynamodb/connection/transactions.py

Co-Authored-By: Ilya Konstantinov <ilya.konstantinov@gmail.com>

hallie committed:
Update pynamodb/models.py

Co-Authored-By: Ilya Konstantinov <ilya.konstantinov@gmail.com>

hallie committed:
Update pynamodb/models.py

Co-Authored-By: Ilya Konstantinov <ilya.konstantinov@gmail.com>

hallie committed:
Update pynamodb/connection/base.pyi

Co-Authored-By: Garrett <garrettheel@users.noreply.github.com>

hallie committed:
addressing comments, adding doc strings

hallie committed:
remove and replace outdated funcitons

hallie committed:
cleaning

hallie committed:
replace connection with fixture

hallie committed:
remove unused ddb_urls

hallie committed:
remove request validation from transaction object

hallie committed:
remove hashing and add test for new error

hallie committed:
using pytest raises

hallie committed:
mispellings and other types

hallie committed:
test frix

hallie committed:
frix weird mymy

hallie committed:
mypy

hallie committed:
another mypy

jpinner-lyft committed:
Allow None to be used to bootstrap condition chaining. (#653)

hallie committed:
restructure api

hallie committed:
cleanup

ikonst committed:
tests: bump to mypy 0.710 (#655)

ikonst committed:
Move pynamodb/tests to tests and exclude them from dist (#628)

ikonst committed:
travis: automatically publish releases to pypi (#654)

hallie committed:
Merge branch 'master' into transact

hallie committed:
add new exceptions

hallie committed:
mypy

hallie committed:
uncancel

hallie committed:
transaction specific operation kwargs

irhkang committed:
Add TTLAttribute and default_as_new (#631)

hallie committed:
Merge https://github.com/pynamodb/PynamoDB into transact

hallie committed:
remove unused

hallie committed:
ignore build

garrettheel committed:
Bump version for v4 beta2 (#636)

garrettheel committed:
Update .travis.yml

hallie committed:
add more tests

hallie committed:
handle future

hallie committed:
Merge branch 'master' into transact

hallie committed:
removing transaction conflict test because its flaky

hallie committed:
bump coverage

hallie committed:
resolve when none

hallie committed:
oops

hallie committed:
addressing feedback

hallie committed:
unit test operation kwargs
